chain/rust/src/byron/utils.rs

```diff
@@ -0,0 +1,496 @@
+            Some(magic) => {
+                if *magic == NetworkInfo::mainnet().protocol_magic() { None } else { protocol_magic }
+            }
+            None => None
+        };
+        AddrAttributes {
+            derivation_path: hdap,
+            stake_distribution: Some(StakeDistribution::BootstrapEra),
+            protocol_magic: adjusted_magic,
+        }
+    }
+    pub fn new_single_key(
+        pubk: &Bip32PublicKey,
+        hdap: Option<HDAddressPayload>,
+        protocol_magic: ProtocolMagic,
+    ) -> Self {
+        AddrAttributes {
+            derivation_path: hdap,
+            stake_distribution: Some(StakeDistribution::new_single_key(StakeholderId::new(pubk))),
+            protocol_magic: Some(protocol_magic),
+        }
+    }
+}
+
+impl AddressId {
+    pub fn new(addr_type: ByronAddrType, spending_data: &SpendingData, attrs: &AddrAttributes) -> Self {
+        // the reason for this unwrap is that we have to dynamically allocate 66 bytes
+        // to serialize 64 bytes in cbor (2 bytes of cbor overhead).
+        let addr_type_uint = addr_type as u32;
+        let buf = cbor!(&(&addr_type_uint, spending_data, attrs))
+            .expect("serialize the AddressId's digest data");
+
+        let hash = Sha3_256::new(&buf);
+        AddressId(*Blake2b224::new(hash.as_ref()).as_hash_bytes())
+    }
+}
+
+impl From<AddressContent> for ByronAddress {
+    fn from(content: AddressContent) -> Self {
+        let content_bytes = content.to_bytes();
+        let crc = super::crc32::crc32(&content_bytes).into();
+        ByronAddress { content, crc }
+    }
+}
+
+impl AddressContent {
+    pub fn hash_and_create(addr_type: ByronAddrType, spending_data: &SpendingData, attributes: AddrAttributes) -> AddressContent {
+        let address_id = AddressId::new(addr_type, spending_data, &attributes);
+        AddressContent::new(
+            address_id, 
+            attributes,
+            addr_type,
+        )
+    }
+
+    // bootstrap era + no hdpayload address
+    pub fn new_redeem(pubkey: PublicKey, protocol_magic: Option<ProtocolMagic>) -> Self {
+        let attributes = AddrAttributes::new_bootstrap_era(None, protocol_magic);
+        let addr_type = ByronAddrType::Redeem;
+        let spending_data = &SpendingData::new_spending_data_redeem(pubkey);
+        
+        AddressContent::hash_and_create(addr_type, &spending_data, attributes)
+    }
+
+    // bootstrap era + no hdpayload address
+    pub fn new_simple(xpub: Bip32PublicKey, protocol_magic: Option<ProtocolMagic>) -> Self {
+        let attributes = AddrAttributes::new_bootstrap_era(None, protocol_magic);
+        let addr_type = ByronAddrType::PublicKey;
+        let spending_data = SpendingData::new_spending_data_pub_key(xpub);
+
+        AddressContent::hash_and_create(addr_type, &spending_data, attributes)
+    }
+
+    /// Do we want to remove this or keep it for people who were using old Byron code?
```

Comment:
I'm okay with either. If we don't need it, it's not hard to replace imo

Reply:
After having run into some type annotations needed in a few places when trying to use the From/Into for addresses, I think we could just keep the to_address()/from_address() things everywhere since it'd also be consistent with wasm (where we can't expose traits) and would help people migrating from CSL/old CML.